src/games/strategy/triplea/ui/AbstractMovePanel.java

```diff
@@ -138,11 +143,67 @@ public final String undoMove(final int moveIndex) {
     return undoMove(moveIndex, false);
   }
 
+  /**
+   * Executes an undo move for any of the units passed in as a parameter.
+   *
+   * "Cannot undo" Error messages are suppressed if any moves cannot be undone
+   * (at least until we come up with a way to deal with "n" reasons for an undo
+   * failure rather than just one)
+   */
+  public void undoMoves(Set<Unit> units) {
+    Set<UndoableMove> movesToUndo = getMovesToUndo(units, getMoveDelegate().getMovesMade());
+
+    if (movesToUndo.size() == 0) {
+      String error = "Could not undo any moves, check that the unit has moved and that you can undo the move normally";
+      JOptionPane.showMessageDialog(getTopLevelAncestor(), error, "Could not undo move", JOptionPane.ERROR_MESSAGE);
+      return;
+    }
+
+    undoMovesInReverseOrder(movesToUndo);
+  }
+
+  private static Set<UndoableMove> getMovesToUndo(Set<Unit> units, List<Object> movesMade) {
+    Set<UndoableMove> movesToUndo = Sets.newHashSet();
+
+    if (movesMade != null) {
+      for (Object undoableMoveObject : movesMade) {
+        if (undoableMoveObject != null) {
+          UndoableMove move = (UndoableMove) undoableMoveObject;
```

Comment:
On line 134 we have:    ` m_undoableMoves = getMoveDelegate().getMovesMade();`

Looking at 'getMovesMade':
```
  /**
   * Get the moves already made
   *
   * @return a list of UndoableMoves or UndoablePlacements
   */
  public abstract List getMovesMade();
```
Hitting the 'u' key for placements simply does not work. It looks like we'd have a bug in existing code if it could be another type.